hadoop-ozone/ozone-manager/src/main/java/org/apache/hadoop/ozone/om/OzoneManagerPrepareState.java

```diff
@@ -0,0 +1,279 @@
+            new String(data, StandardCharsets.UTF_8));
+      } catch (NumberFormatException e) {
+        LOG.error("Failed to parse log index from prepare marker file {} " +
+            "while restoring OM.", prepareMarkerFile.getAbsolutePath());
+        prepareIndexRead = false;
+      }
+    } else {
+      // No marker file found.
+      prepareIndexRead = false;
+    }
+
+    boolean prepareRestored = false;
+    if (prepareIndexRead) {
+      if (prepareMarkerIndex != expectedPrepareIndex) {
+        LOG.error("Failed to restore OM prepare state, because the expected " +
+            "prepare index {} does not match the index {} written to the " +
+            "marker file.", expectedPrepareIndex, prepareMarkerIndex);
+      } else {
+        // Prepare state can only be restored if we read the expected index
+        // from the marker file.
+        prepareRestored = true;
+      }
+    }
+
+    if (prepareRestored) {
+      enablePrepareGate();
+      // Do not rewrite the marker file, since we verified it already exists.
+      finishPrepare(prepareMarkerIndex, false);
+    } else {
+      // If the potentially faulty marker file cannot be deleted,
+      // propagate the IOException.
+      // If there is no marker file, this call sets the in memory state only.
+      cancelPrepare();
+    }
+
+    return status;
+  }
+
+  /**
+   * If the prepare gate is enabled, always returns true.
+   * If the prepare gate is disabled, returns true only if {@code
+   * requestType} is {@code Prepare} or {@code CancelPrepare}. Returns false
+   * otherwise.
+   */
+  public synchronized boolean requestAllowed(Type requestType) {
+    boolean requestAllowed = true;
+
+    if (prepareGateEnabled) {
+      // TODO: Also return true for cancel prepare when it is implemented.
+      requestAllowed = (requestType == Type.Prepare);
+    }
+
+    return requestAllowed;
+  }
+
+  /**
+   * @return the current log index and status of preparation.
+   * Both fields are returned together to provide a consistent view of the
+   * state, which would not be guaranteed if they had to be retrieved through
+   * separate getters.
+   */
+  public synchronized State getState() {
+    return new State(prepareIndex, status);
+  }
+
+  /**
+   * Creates a prepare marker file inside {@code metadataDir} which contains
+   * the log index {@code index}. If a marker file already exists, it will be
+   * overwritten.
+   */
+  private void writePrepareMarkerFile(long index) throws IOException {
+    File markerFile = getPrepareMarkerFile();
+    File parentDir = markerFile.getParentFile();
+    if (!parentDir.mkdirs()) {
+      throw new IOException("Failed to create necessary directories in " +
+          "marker file path: " + parentDir);
+    }
+
+    try(FileOutputStream stream = new FileOutputStream(markerFile)) {
+      stream.write(Long.toString(index).getBytes(StandardCharsets.UTF_8));
+    }
+  }
+
+  private void deletePrepareMarkerFile()
```

Comment:
Can we add a log line when we create & delete the marker file? It will help us with troubleshooting issues with preparation.